reana_server/complexity.py

```diff
@@ -0,0 +1,354 @@
+
+    def get_bin(memory):
+        for b in range(bins):
+            if memory in range(bin_size * b, bin_size * (b + 1)):
+                return b
+        return bins - 1
+
+    # Summing up parrarel jobs times memory to determine the workflow complexity
+    memory = int(reduce(lambda sum, item: sum + item[0] * item[1], complexity, 0))
+    bins = MQ_MAX_PRIORITY
+    total_cluster_memory = NodesStatus().get_total_memory()
+    bin_size = int(total_cluster_memory / bins)
+    priority = bins - get_bin(memory)
+    return priority
+
+
+def estimate_complexity(workflow_type, reana_yaml):
+    """Estimate complexity in REANA workflow.
+
+    :param workflow_type: A supported workflow specification type.
+    :param reana_yaml: REANA YAML specification.
+    """
+
+    def build_estimator(workflow_type, reana_yaml):
+        if workflow_type == "serial":
+            return SerialComplexityEstimator(reana_yaml)
+        elif workflow_type == "yadage":
+            return YadageComplexityEstimator(reana_yaml)
+        elif workflow_type == "cwl":
+            return CWLComplexityEstimator(reana_yaml)
+        else:
+            raise Exception(
+                "Workflow type '{0}' is not supported".format(workflow_type)
+            )
+
+    estimator = build_estimator(workflow_type, reana_yaml)
+    try:
+        complexity = estimator.estimate_complexity()
+    except Exception:
+        return []
+    return complexity
+
+
+class ComplexityEstimatorBase:
+    """REANA workflow complexity estimator base class."""
+
+    def __init__(self, reana_yaml):
+        """Estimate complexity in REANA workflow.
+
+        :param reana_yaml: REANA YAML specification.
+        :param initial_step: initial workflow execution step.
+        """
+        self.reana_yaml = reana_yaml
+        self.specification = reana_yaml.get("workflow", {}).get("specification", {})
+        self.input_params = reana_yaml.get("inputs", {}).get("parameters", {})
+
+    def parse_specification(self, initial_step):
+        """Parse REANA workflow specification tree."""
+        raise NotImplementedError
+
+    def estimate_complexity(self, initial_step="init"):
+        """Estimate complexity in parsed REANA workflow tree."""
+        steps = self.parse_specification(initial_step)
+        return self._calculate_complexity(steps)
+
+    def _calculate_complexity(self, steps):
+        """Calculate complexity in parsed REANA workflow tree."""
+        complexity = []
+        for step in steps.values():
+            complexity += step["complexity"]
+        return complexity
+
+    def _get_number_of_jobs(self, step):
+        """Get number of jobs based on compute backend."""
+        backend = step.get("compute_backend")
+        if backend and backend != "kubernetes":
+            return 0
+        return 1
+
+    def _get_memory_limit(self, step):
+        """Get memory limit value."""
+        default_memory_limit = os.getenv("REANA_KUBERNETES_JOBS_MEMORY_LIMIT", "8Gi")
```

Comment:
I got some problems due to this default limit. My local cluster has 6Gi, and any workflow I run without `kubernetes_memory_limit` keeps being queued forever as `check_memory_availability` is `False`.

This means that any cluster with less than `8Gi` won't be able to schedule workflows with the default configuration. I think this can be tricky for small instances, or multi-node instances with _small_ (<8Gi) nodes.

In addition, I'd move this default limit to the `config.py` and import `REANA_KUBERNETES_JOBS_MEMORY_LIMIT` from `config.py` as well, as r-w-c does. Alternatively, as this is being used in both r-server and r-w-e, we can put it in reana-commons and import it from there.

Reply:
Great that you spotted this one! As a fix I moved `REANA_KUBERNETES_JOBS_MEMORY_LIMIT` limit to `config.py` in r-server. Since it have a default value, I didn't want to put it in r-commons.

Also, reduced the default value from `8Gi` to `4Gi` to solve `check_memory_availability` issue. It's not ideal, but should work in most of the cases